docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+# User-Defined Type Functions
+
+## Summary
+
+The new type solver for Luau supports a limited number of built-in type functions that enable developers to be expressive with their code. To further enhance this expressivity and the flexibility of the type system, this RFC proposes the design and implementation of user-defined type functions, a feature that will allow developers to define custom type functions.
+
+## Motivation
+
+The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
+- Enable more precise type definitions that can capture complex relationships and constraints
+- Facilitate the creation of reusable and composable libraries
+- Enhance code safety and maintainability with better type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.
+
+## Design
+
+Type functions wil be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
+
+For instance, the `rawget` type function can be written as:
+```luau
+type function rawget(tbl, prop)
+    return tbl.getProps()[prop.getValue()]
+
+    -- returning nothing / nil means that this type function will fail to reduce
+end
+
+type Person = {
+    name: string,
+    age: number
+}
+
+type ty = rawget<Person, "name"> -- ty = string
+```
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
+
+<details><summary>lType library methods (dropdown)</summary>
+Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).
+
+#### Any
+
+| Function Declaration | Return Type | Description |
```

Comment:
I don't see anything about union types or intersection types.

Reply:
Or generics, or type function instances.

Reply:
Negation types.

Reply:
Error suppressing types?

Reply:
Type packs as well!

Reply:
I think we'll probably consider type pack functions to be out of scope for the initial version, but if people have strong use cases for them, we can certainly consider it now.

Reply:
> I don't see anything about union types or intersection types.
> Negation types.
> Error suppressing types?

Added!

> Or generics, or type function instances.

I didn't really see an use case for this, so I didn't bring it up during my revision. do you think it's a must have for at least the initial version?

Reply:
> Type packs as well! typeFunc<(A, B), (C, D)>
And variadic types. typeFunc<...A>

Following Aaron's comment, the initial version only includes the ability to add type packs/variadic types to an instance of a function type, but doesn't have anything about the type functions themselves taking in type packs/variadic types